Fix issues with multiline generic type parameters

josh-degraw opened the pull request:
Closes #2706, from Amplifying F# session

nojaf commented:
Thanks for following up on this. I need to play with this some more and we may want to have the conversation over at the style guide as well.
Once all is settled, we should target 6.1 for this.

nojaf commented on src/Fantomas.Core/CodePrinter.fs:
```diff
@@ -3037,7 +3045,8 @@ let genType (t: Type) =
         +> genSingleTextNode node.LessThen
         +> addExtraSpace
-        +> col sepComma node.Arguments genType
+        +> leadingExpressionIsMultiline (colGenericTypeParameters node.Arguments) (fun isMultiline ->
```
We should add some tests where the `<` and `>` contain trivia.
Cases like:
```fsharp
type X =
    Teq< //
        int
     //
     >
```
won't be covered right now.

nojaf replied:
Also please add a code comment on why we are doing this in the first place.

nojaf commented on src/Fantomas.Core/CodePrinter.fs:
```diff
@@ -3037,7 +3045,8 @@ let genType (t: Type) =
         +> addExtraSpace
-        +> col sepComma node.Arguments genType
+        +> leadingExpressionIsMultiline (colGenericTypeParameters node.Arguments) (fun isMultiline ->
+            if isMultiline then !- " " else sepNone)
```
Use helper `onlyIf` here.

nojaf commented on src/Fantomas.Core/CodePrinter.fs:
```diff
@@ -2191,13 +2191,21 @@ let genKeepIdent (startNode: Node) (e: Expr) ctx =
-            match t with
-            | Type.Var n when idx = 0 && n.Text.StartsWith("^") -> sepSpace
-            | _ -> sepNone
+    let short sep =
```
Maybe rename to `genParameters` instead of `short`.
We could have a separate `let short = genParameters sepComma`.

nojaf commented on src/Fantomas.Core.Tests/TypeAnnotationTests.fs:
```diff
@@ -122,11 +127,79 @@ type CancellableTaskResultBuilderBase with
+"""
+
+[<Test>]
+let `` Aligned bracket style in anonymous record is respected for multiple types, #2706`` () =
```
We should have the same test for the other styles.

nojaf reviewed (changes requested):
Hi @josh-degraw, many thanks again for taking the lead in this during https://amplifying-fsharp.github.io/sessions/2023/04/14/

I think the change is sensible but this does change the style and so I would like to see this covered by https://learn.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting#formatting-explicit-generic-type-arguments-and-constraints

And also to have a link for people who will open an issue claiming the additional space is a bug. (Trust me, it will be reported 🙈).

Could you open an issue in https://github.com/fsharp/fslang-design/issues so we can further discuss these implications?

And please also retarget this branch to `v6.1`.

Thanks!

josh-degraw commented:
Sorry it's taken me so long to finally get back around to trying to finish this up haha.

I'm not sure if something has changed on the parser level since I originally worked on this, but a test that previously succeeded is now failing and I'm having trouble figuring out the best way to address it.

I extracted the relevant bits into its own test but essentially this kind of expression (which is how my PR originally formatted this in the updated tests here):
```fs
let bv =
    unbox<
        Foo<
            'innerContextLongLongLong,
            'bb -> 'b
            >
    >
        bf
```

Is no longer an idempotent operation. It seems like if I increase the indent of the final `>` when it's output in `genExpr` under `Expr.TypeApp` so it looks like this:
```fs
let bv =
    unbox<
        Foo<
            'innerContextLongLongLong,
            'bb -> 'b
            >
     >
        bf
```

it becomes idempotent again, but since this worked before I rebased this PR I'm not sure if this is a bug in the AST parsing layer or just in my code here.

nojaf commented:
@josh-degraw thanks for picking this up again! Much appreciated!

I'm afraid this one kinda sucks:

![image](https://github.com/fsprojects/fantomas/assets/2621499/ec68ffa1-3452-4c95-a3a1-fa17c23b3f18)

The `unbox<...> bf` no longer is a function application after formatting.

So, instead of `a b` it becomes `a ; b` conceptually.

I'm not sure what we need to do to keep this as a function application.

josh-degraw commented:
> @josh-degraw thanks for picking this up again! Much appreciated!
> 
> I'm afraid this one kinda sucks:
> 
> ![image](https://github.com/fsprojects/fantomas/assets/2621499/ec68ffa1-3452-4c95-a3a1-fa17c23b3f18)
> 
> The `unbox<...> bf` no longer is a function application after formatting.
> 
> So, instead of `a b` it becomes `a ; b` conceptually.
> 
> I'm not sure what we need to do to keep this as a function application.

Yeah like I mentioned it seems like just increasing the indent for the closing angle bracket at least fixes the parser issue, but that's still valid code with the additional indent right? If not I guess we'd have to carve out an exemption for this scenario or wrap in parentheses or something

nojaf commented:
Ok, consider this example:

```fsharp
foo<
    'bar
        -> int
>
    barry
```

This looks like it is a `SynExpr.App(foo<...>, barry)` but actually it is already is two top level `SynModuleDecl.Expr`.

Adding a space before `>` makes it into a `SynModuleDecl.Expr (SynExpr.App ...)`.
So, I think our hand is forced here to just move it one space further.

nojaf commented:
Added a workaround of `let startColumn = ctx.Column + 1`

josh-degraw commented:
I tried that but that messes with a bunch of other tests, wouldn't we want to only do this in this specific situation?

nojaf commented:
Hmm, I think the safest thing to do would be to have it all the time.
We essentially do the same thing with `SynType` I think.

Digging into the exact case of `SynExpr.App(SynExpr.TypeApp, ...)` might be quite challenging and not really worth it.

josh-degraw commented:
Fair enough. Looks like most of the tests that fail with that change were added in this PR anyway haha. I'll update the tests to reflect this

nojaf commented:
@josh-degraw took a quick peek if it is possible. It seems less intrusive at first glance to remove the space.

This of course can still go wrong in things like `Expr.AppLongIdentAndSingleParenArg`, `Expr.AppSingleParenArg`, `Expr.AppWithLambda`, `Expr.NestedIndexWithoutDot`, `EndsWithDualListApp`, `EndsWithSingleListApp`, ...

We should check what happens in those scenarios. If we need to update all these individually I don't think it is worth it.

josh-degraw commented on src/Fantomas.Core.Tests/DallasTests.fs:
```diff
@@ -1596,11 +1596,14 @@ let autoCompleteItems: cmap<DeclName, DeclarationListItem * Position * string<Lo
-      (Position -> option<string>) *
-      FSharp.Compiler.Syntax.ParsedInput> =
+    : cmap<
+          DeclName,
```
Do you still consider this a problem or are we okay with this indent level?

nojaf replied:
Hmm, I think I can live with this.
This sample screams "use type alias" anyway.

josh-degraw commented:
> This of course can still go wrong in things like Expr.AppLongIdentAndSingleParenArg, Expr.AppSingleParenArg, Expr.AppWithLambda, Expr.NestedIndexWithoutDot, EndsWithDualListApp, EndsWithSingleListApp, ...

Can you maybe help me know in what situations these nodes would apply here? I'm not quite sure exactly how to test for these just going off of the node names themselves (not quite there yet haha..)

nojaf commented:
> Can you maybe help me know in what situations these nodes would apply here?

Sure thing.

[SynExpr.App](https://fsprojects.github.io/fantomas/reference/fantomas-fcs-syntax-synexpr.html#App) has historically been a tricky one to format. Because it really depends on what expressions it has before we decide how we can format it.

Some [examples](https://fsprojects.github.io/fantomas-tools/#/oak?data=N4KABGBEDGD2AmBTSAuKB6dYCiAPADgE4CC%2B%2BAygJYB2A5gDaIAKAhoYtcYbQHIKIAdQgOoAxWLAAUQkRAiYZ1OWABGbRQEpFizDgIkyAdUoAXABYAZFgFsV8FnySK1hSQDMArktxgAtAD4wXC1hah0sPCJSfEdBUJYwAG0wAAl2NzABSABnWGtEX3oaAGssgF0kxWVsk0JMmFhqEw4TLMUy7VDdSMIASWo3Slxo2MzQn18wAE9OkW79chtEfsHhsmzYxR8AKmmwXYAvSAAacCgkQepEbNQoE7PISmzRbMpbtxZ6bMRTiEfsgDihBY%2BDMADVKIgAO7vT7fEAAXyAA):
```fsharp
// ExprAppSingleParenArgNode
Foo(
    //
    bar
)

// ExprAppWithLambdaNode
bar(fun x -> x)

// ExprAppNode
a [ Href "some-link"] [
    str "content"
]

// ExprInfixAppNode 
x - y

// ExprSameInfixAppsNode
x * y * z
```

![image](https://github.com/fsprojects/fantomas/assets/2621499/2365d6aa-f4fd-4bf0-99fb-ffebf6025c94)

All of these expressions are `SynExpr.App`, but we transform them into different `Oak` shapes in `ASTTransformer`.

This is sometimes very convenient but at the same has its drawbacks.

A clear pro: sometimes you really need to deal with these shapes very carefully and it helps to have separation.

A clear con: the is no single entry `genApp` that deals with all types of applications. It can be a bit scattered.

A lot of what we have today is somewhat historical as well. Infix applications are a good example of that. Almost all of the code pre-dates the style guide and it is just what we have today.

We split the applications into different Oak nodes in:
https://github.com/fsprojects/fantomas/blob/21bae08e87b63b2f9824ec19d4aa326d16309075/src/Fantomas.Core/ASTTransformer.fs#L1184-L1315

So, to answer your question, that would be the list you need to check.

On a side note, some of these nodes probably shouldn't exist because they are too focused on what we do with them inside `CodePrinter`. For this reason, you also have `EndsWithDualListApp` as an active pattern instead of a separate node. It is hard to really draw the line there.

Anyway, you should check what happens if any of these nodes can be used with `ExprTypeAppNode` and what happens if they do with a specific case.

I hope this helps a little.

nojaf commented on src/Fantomas.Core.Tests/TypeAnnotationTests.fs:
```diff
@@ -266,25 +262,27 @@ let private asJson (arm: IArmResource) =
        >
 """
 
+let forcedLongDefnConfig =
```
I'm not a fan of this configuration.
The only thing you are trying to achieve is to spread things over multiple lines, using `MaxLineLength = 30`. All the other settings are irrelevant and misleading to readers.

josh-degraw replied:
Yeah, fully agree, this latest commit was mainly a WIP

nojaf commented on src/Fantomas.Core.Tests/TypeAnnotationTests.fs:
```diff
@@ -299,82 +297,211 @@ let bv =
-    |}>
-"""     
-        config
+let ``Multiline type argument with AppLongIdentAndSingleParenArg`` () =
```
Nit: lowercase unit test names.

nojaf commented on src/Fantomas.Core.Tests/TypeAnnotationTests.fs:
```diff
@@ -262,23 +262,11 @@ let private asJson (arm: IArmResource) =
 [<Test>]
 let ``type application including nested multiline function type`` () =
-    forcedLongDefnConfig
+    { config with
```
Nitpick but this doesn't comply with the [pull request ground rules](https://fsprojects.github.io/fantomas/docs/contributors/Pull%20request%20ground%20rules.html).

I don't really see a compelling reason to deviate from the standard unit template.
All our tests are the same, I'm not a fan of mixing that up in this PR.

nojaf commented:
Be aware that in its current form this is not working out for every `SynExpr.App` combination.
I did a quick test with:

```fsharp
[<Test>]
let ``dual list application`` () =
    formatSourceString
        """
unbox<Foo<'innerContextLongLongLong,'bb -> 'b>> [
    ClassName "meh"
] [
    str "foo"
]
"""
        { config with
            MaxLineLength = 12
            MultilineBracketStyle = Stroustrup }
    |> prepend newline
    |> should
        equal
        """
unbox<
    Foo<
        'innerContextLongLongLong,
        'bb
            -> 'b
     >
> [
    ClassName 
        "meh"
] [
    str 
        "foo"
]
"""
```

josh-degraw commented:
> Be aware that in its current form this is not working out for every `SynExpr.App` combination.

Yeah I know, I'm still having a little trouble pinpointing the spot to fix some of these

josh-degraw commented:
> Don't hesitate to reach out if you wanna pair on this sometime!

That might be helpful. I felt like a few I found right where I thought I needed to make changes to fix some of these issues but it didn't work haha

nojaf reviewed (approved):
Alright, I think this is worth merging in and release as an alpha.
Thank you for your endurance here @josh-degraw!
